File: srclab-common/src/main/java/xyz/srclab/common/reflect/ClassKit.java
package xyz.srclab.common.reflect;

import xyz.srclab.annotation.Nullable;
import xyz.srclab.common.base.As;
import xyz.srclab.common.base.Check;
import xyz.srclab.common.base.Loader;
import xyz.srclab.common.invoke.ConstructorInvoker;
import xyz.srclab.common.lang.finder.Finder;

import java.lang.reflect.Constructor;

/**
 * @author sunqian
 */
public class ClassKit {

    public static <T> T newInstance(String className) {
        @Nullable Class<?> cls = Loader.findClass(className);
        Check.checkArgument(cls != null, "Class not found: " + className);
        return newInstance(cls);
    }

    public static <T> T newInstance(String className, ClassLoader classLoader) {
        @Nullable Class<?> cls = Loader.findClass(className, classLoader);
        Check.checkArgument(cls != null, "Class not found: " + className);
        return newInstance(cls);
    }

    public static <T> T newInstance(Class<?> cls) {
        @Nullable Constructor<?> constructor = ConstructorKit.getConstructor(cls);
        Check.checkArgument(constructor != null, "Constructor not found: " + cls);
        return As.notNull(ConstructorInvoker.of(constructor).invoke());
    }

    public static <T> T newInstance(Class<?> cls, Class<?>[] parameterTypes, Object[] arguments) {
        @Nullable Constructor<?> constructor = ConstructorKit.getConstructor(cls, parameterTypes);
        Check.checkArgument(constructor != null, "Constructor not found: " + cls);
        return As.notNull(ConstructorInvoker.of(constructor).invoke(arguments));
    }

    public static Class<?> toWrapper(Class<?> primitive) {
        @Nullable Class<?> wrapper = WrapperFinder.find(primitive);
        return wrapper == null ? primitive : wrapper;
    }

    private static final class WrapperFinder {

        private static final Finder<Class<?>, Class<?>> finder = Finder.pairHashFinder(WrapperClasses.TABLE);

        @Nullable
        public static Class<?> find(Class<?> primitive) {
            return finder.find(primitive);
        }

        private static final class WrapperClasses {

            private static final Class<?>[] TABLE = {
                    boolean.class, Boolean.class,
                    byte.class, Byte.class,
                    short.class, Short.class,
                    char.class, Character.class,
                    int.class, Integer.class,
                    long.class, Long.class,
                    float.class, Float.class,
                    double.class, Double.class,
                    void.class, Void.class,
            };
        }
    }
}

File: srclab-common/src/main/kotlin/xyz/srclab/common/reflect/ClassKit.kt
@file:JvmName("ClassKit")
@file:JvmMultifileClass

package xyz.srclab.common.reflect

import xyz.srclab.common.base.Current
import xyz.srclab.common.base.asAny

@JvmOverloads
fun <T> CharSequence.findClass(classLoader: ClassLoader = Current.classLoader): Class<T>? {
    return try {
        Class.forName(this.toString(), true, classLoader)
    } catch (e: ClassNotFoundException) {
        null
    }.asAny()
}

@JvmOverloads
fun <T> CharSequence.findClassToInstance(classLoader: ClassLoader = Current.classLoader): T? {
    return this.findClassToInstance(classLoader, emptyArray(), emptyArray())
}

@JvmOverloads
fun <T> CharSequence.findClassToInstance(
    classLoader: ClassLoader = Current.classLoader,
    parameterTypes: Array<out Class<*>>,
    args: Array<out Any?>
): T? {
    val cls: Class<T>? = this.findClass(classLoader)
    if (cls === null) {
        return null
    }
    return cls.toInstance(parameterTypes, args)
}

fun <T> Class<*>.toInstance(): T {
    return this.toInstance(emptyArray(), emptyArray())
}

fun <T> Class<*>.toInstance(parameterTypes: Array<out Class<*>>, args: Array<out Any?>): T {
    val constructor = this.findConstructor(*parameterTypes)
    if (constructor === null) {
        throw IllegalArgumentException("Class constructor(${parameterTypes.contentToString()}) not found: $this")
    }
    return constructor.newInstance(*args).asAny()
}

fun Class<*>.toWrapperClass(): Class<*> {
    return when (this) {
        Boolean::class.javaPrimitiveType -> Boolean::class.java
        Byte::class.javaPrimitiveType -> Byte::class.java
        Short::class.javaPrimitiveType -> Short::class.java
        Char::class.javaPrimitiveType -> Char::class.java
        Int::class.javaPrimitiveType -> Int::class.java
        Long::class.javaPrimitiveType -> Long::class.java
        Float::class.javaPrimitiveType -> Float::class.java
        Double::class.javaPrimitiveType -> Double::class.java
        Void::class.javaPrimitiveType -> Void::class.java
        else -> this
    }
}